### PR TITLE
Streaming store-gateway: chunks grouping

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
### Context 

See https://github.com/grafana/mimir/issues/3939

### What this PR does

__TL;DR:__ this PR solves two problems
* because we try to save number of API calls to the object storage we also end up fetching more data from the chunks cache because we apply the same strategy of joining nearby byte ranges
* because we estimate the length of every single chunk as 16KB, we end up fetching much more than we need

__Longer version__

* changes `bucketIndexLoadedSeries` to infer the chunks length from the index on a series
	* _How?_ we know that chunk ref is delta encoded _and_ since chunks of a single series are contiguous in the segment file. For chunks within the same segment file, the delta in the series bytes is effectively the length of the previous chunk.
* changes `loadingSeriesChunkRefsSetIterator` to partition chunks into groups instead of passing them directly to the `loadingSeriesChunksSetIterator`
	* each group contains chunks only from a single segment file, since that's how we fetch them from the bucket
* manage our own chunks cache instead of relying on a `CachingBucket`
	* `loadingSeriesChunksSetIterator` caches each individual group
	* I've added a new package `github.com/grafana/mimir/pkg/storegateway/chunkscache` that has a naive inmemory implementation and a memcached implementation
	* the cache key is 
		* block ULID & tenant ID
		* the ref of the first chunk in the group 
		* the number of chunks in the group
	* the cached value is the concatenated chunk bytes (`[len, encoding, data, crc32]`) for each chunk (the same way they are stored in the segment files)

### Notes to reviewers

* There are some missing tests marked with TODOs, I will address those
* There is at least one bug because testing this in a dev cluster has roughly 0.5% incorrect results
* This will be in draft until the above are fixed and we're more settled on the general design.
* I am also considering breaking up this PR into separate PRs to merge into `main`:
	* introducing the chunks cache package
	* the changes to `bucketIndexLoadedSeries` (and small parts of `loadingSeriesChunkRefsSetIterator`)
	* the changes to `loadingSeriesChunkRefsSetIterator` (and small parts of `loadingSeriesChunksSetIterator`)
	* the changes to `loadingSeriesChunksSetIterator`

### Further work

Probably not part of this PR

* Due to how memcached `SET` is implemented right now, caching the group is inefficient because we copy each byte slice we want to cache. This is because the `SET`s are asynchronous, so there is a race condition between sending the byte slice to memcached and releasing the pooled byte slice for the chunk
* For when we don't know the size of a chunk group we still use the 16KB estimation, I think we can improve this and reduce this estimation (see https://github.com/grafana/mimir/issues/3939#issuecomment-1381515942)

### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/3939

### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by dimitarvdimitrov** [comment]
Here are some preliminary performance results - this is from querying few series over a ~72h time range

* zone-a is running streaming + mmap-less on r220
* zone-b is running streaming + mmap-less on this PR
* zone-c is running the default implementation

Screenshot 2023-01-16 at 22 43 57
Screenshot 2023-01-16 at 22 44 25

File: pkg/storegateway/series_chunks.go [hunk 1]
```diff
@@ -335,36 +420,208 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
...
+			return false, int64(totalChunkLen), nil
+		}
+		c := rawChunk(gBytes[n : n+1+int(chunkDataLen)])
+		if cEnc := c.Encoding(); cEnc != chunkenc.EncXOR {
+			return false, 0, fmt.Errorf("encoding (%d, %s) isn't XOR, don't know what to do ", cEnc, cEnc.String())
+		}
+		chunks[i].Raw.Type = storepb.Chunk_XOR
+		chunks[i].Raw.Data = c.Bytes()
+		// We ignore the crc32 because we assume that the chunk didn't get corrupted.
+		// TODO maybe check the crc for every 1 in 100 chunks? 1 in 1000?
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/series_chunks.go:603]: Thoughts? I'm not sure what guarantees us that the bytes we get from memcached and the bucket will not be corrupted. Probably TCP?

**Comment by 56quarters** [comment]
> ### Further work
> 
> Probably not part of this PR
> 
>  * Due to how memcached `SET` is implemented right now, caching the group is inefficient because we copy each byte slice we want to cache. This is because the `SET`s are asynchronous, so there is a race condition between sending the byte slice to memcached and releasing the pooled byte slice for the chunk

As caching is currently written, it'll be difficult to use memory pooling for `SET`s. Since the interface for the cache abstraction in dskit is `Store(map[string][]byte)`, it loops over the map doing a `Set(string, []byte)` for each key-value which is put into a job queue. There are multiple workers pulling from the job queue so it's difficult to determine when it's safe to free a chunk of memory used in a `Store(map[string][]byte)` call. 

There are three options that I can think of for this offhand:

* Change how the cache abstraction in dskit works so that we could easily determine when to call `Allocator.Release()`. We wouldn't split a set of multiple keys into separate tasks, just us a single task which loops over the values, setting them, and calls `.Release()` at the end. This still lets callers of the cache methods return immediately but makes things deterministic enough for us to implement pooling.
* Don't really change how any of the cache abstraction in dskit works and use a `sync.WaitGroup` and goroutine to call `Allocator.Release()` after the last `.Set(string, []byte)` is run. I don't particularly like this option since it feels prone to leak resources.
* Do nothing - maybe things are good enough that we should leave it as-is for now

File: pkg/storegateway/bucket_index_reader.go [hunk 2]
```diff
@@ -581,6 +583,63 @@ func (l *bucketIndexLoadedSeries) addSeries(ref storage.SeriesRef, data []byte)
 	l.seriesMx.Unlock()
 }
 
+// chunksGroup contains chunks from the same block and the same segment file. They are ordered by their minTime
+type chunksGroup struct {
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/bucket_index_reader.go:587]: this should probably be `seriesChunksGroup`. i can rename

**Comment by dimitarvdimitrov** [comment]
> Since the interface for the cache abstraction in dskit is `Store(map[string][]byte)`,

I think we're using the `RemoteCacheClient.SetAsync` method to set our items ([code](https://github.com/grafana/mimir/blob/2393f181df59af50c48484834473886375e2325e/vendor/github.com/grafana/dskit/cache/memcached_client.go#L238-L247)). It takes only a single key. Can we not take an `Allocator` in the same function and put back the byte slice when `SetAsync` is done?

I couldn't find where we use `Cache` instead of `RemoteCacheClient`

**Comment by 56quarters** [comment]
> > Since the interface for the cache abstraction in dskit is `Store(map[string][]byte)`,
> 
> I think we're using the `RemoteCacheClient.SetAsync` method to set our items ([code in dskit](https://github.com/grafana/mimir/blob/2393f181df59af50c48484834473886375e2325e/vendor/github.com/grafana/dskit/cache/memcached_client.go#L238-L247), [code in the store-gateway](https://github.com/grafana/mimir/blob/a95cb4b1e59452bb92a3d45b55218ecc4c354f4b/pkg/storegateway/chunkscache/memcached.go#L94)). It takes only a single key. Can we not take an `Allocator` in the same function and put back the byte slice when `SetAsync` is done?

I was mistaken, we are indeed using `RemoteCacheClient` instead of `Cache` (we need to clean up the abstractions we offer in dskit but that's another problem). Either way, the code is async so it's difficult to tell when the memory is safe to free.

We could add an `Allocator` to the `Store` or `SetAsync` methods but the idea behind adding the `Allocator` to the get methods was being able to free everything associated with a request at once - the `Put(b *[]byte)` method is a no-op in Mimir as it stands today and the `Allocator` was expected to actually be a `SafeSlabPool`.

So we'd either have to start using something that's not `SafeSlabPool` for the memory used when storing results in cache or add cache methods that block when doing stores.

I think pooling memory for stores is probably not worth solving right now.

File: pkg/storegateway/chunkscache/cache.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+type Range struct {
+	BlockID   ulid.ULID
+	Start     chunks.ChunkRef
+	NumChunks int // We are certain about the number of chunks, not about their total size
+}
+
+type ChunksCache interface {
```
Comment (56quarters) [inline pkg/storegateway/chunkscache/cache.go:20]: If `ranges` is always required and `bytesPool` is sometimes `nil` my preference would be to make `bytesPool` the last argument

File: pkg/storegateway/chunkscache/memcached.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+const (
+	memcachedDefaultTTL = 7 * 24 * time.Hour
+)
+
+// MemcachedChunksCache is a memcached-based index cache.
+type MemcachedChunksCache struct {
+	logger    log.Logger
+	memcached cache.MemcachedClient
```
Comment (56quarters) [inline pkg/storegateway/chunkscache/memcached.go:29]: `cache.MemcachedClient` is an alias for `cache.RemoteCacheClient`.

File: pkg/storegateway/chunkscache/memcached.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/thanos-io/thanos/blob/main/pkg/store/cache/memcached.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Thanos Authors.
```
Comment (56quarters) [inline pkg/storegateway/chunkscache/memcached.go:4]: This is new code, no?

File: pkg/storegateway/series_chunks.go [hunk 6]
```diff
@@ -320,9 +364,50 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 	// This data structure doesn't retain the seriesChunkRefsSet so it can be released once done.
 	defer nextUnloaded.release()
 
+	c.chunkReaders.reset()
+	groupCount := 0
+	for _, s := range nextUnloaded.series {
+		groupCount += len(s.groups)
+	}
+	loadedGroups := make([][]byte, groupCount)
```
Comment (dimitarvdimitrov) [inline pkg/storegateway/series_chunks.go:372]: keeping track of this slice along with the `var currentGroup int` counter and using `nextUnloaded.series` makes all this a bit clunky. I will add an intermediate type

File: pkg/storegateway/chunkscache/memcached.go (hunk 4 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/chunkscache/memcached.go:29]: are you saying we should use `cache.RemoteCacheClient` instead? Why do we even have two 😄 ?

File: pkg/storegateway/chunkscache/memcached.go (hunk 5 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/chunkscache/memcached.go:4]: meh, kind of. I copied it from pkg/storegateway/indexcache/memcached.go   which has this header 

https://github.com/grafana/mimir/blob/65bde07f57d1b3cdf7ffa54d5af9b244e2fca4dd/pkg/storegateway/indexcache/memcached.go#L1-L4

File: pkg/storegateway/chunkscache/cache.go (hunk 3 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/chunkscache/cache.go:20]: i agree. previously ranges was variadic, so i couldn't do that. But when i changed it to a regular parameter, i didn't change the order too. refactored

File: pkg/storage/tsdb/index_cache.go [hunk 7]
```diff
@@ -119,3 +121,18 @@ func newMemcachedIndexCache(cfg cache.MemcachedConfig, logger log.Logger, regist
...
 }
+
+func NewChunksCache(cfg ChunksCacheConfig, logger log.Logger, reg prometheus.Registerer) (chunkscache.ChunksCache, error) {
+	switch cfg.Backend {
+	case "":
+		return nil, nil
+	case cache.BackendMemcached:
+		client, err := cache.NewMemcachedClientWithConfig(logger, "chunks-cache", cfg.Memcached.ToMemcachedClientConfig(), prometheus.WrapRegistererWithPrefix("thanos_", reg))
+		if err != nil {
+			return nil, errors.Wrap(err, "create index cache memcached client")
```
Comment (pracucci) [inline pkg/storage/tsdb/index_cache.go:132]: [nit]
```suggestion
			return nil, errors.Wrap(err, "create chunks cache memcached client")
```

File: pkg/storegateway/series_refs.go [hunk 8]
```diff
@@ -106,14 +106,14 @@ func (b seriesChunkRefsSet) release() {
 // seriesChunkRefs holds a series with a list of chunk references.
 type seriesChunkRefs struct {
 	lset   labels.Labels
-	chunks []seriesChunkRef
+	groups []seriesChunkRefsGroup
```
Comment (pracucci) [inline pkg/storegateway/series_refs.go:109]: [nit] I would clarify it's a group of chunk refs.

```suggestion
	chunkGroups []seriesChunkRefsGroup
```

File: pkg/storegateway/series_refs.go [hunk 9]
```diff
@@ -805,7 +838,7 @@ func (s *loadingSeriesChunkRefsSetIterator) Err() error {
 }
 
 func (s *loadingSeriesChunkRefsSetIterator) loadSeriesForTime(ref storage.SeriesRef, loadedSeries *bucketIndexLoadedSeries, stats *queryStats) (labels.Labels, []seriesChunkRef, error) {
```
Comment (pracucci) [inline pkg/storegateway/series_refs.go:840]: [nit] `loadSeriesForTime` should be renamed because it's not `ForTime` anymore.

File: pkg/storegateway/series_refs.go [hunk 10]
```diff
@@ -766,18 +766,25 @@ func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
...
-		if lset.Len() == 0 {
-			// No matching chunks for this time duration, skip series
+		if len(lset) == 0 {
+			// This was a series without any chunks; we skip it
 			continue
 		}
-
+		var groups []seriesChunkRefsGroup
+		if !s.skipChunks {
+			groups = partitionChunks(chks, s.blockID)
```
Comment (pracucci) [inline pkg/storegateway/series_refs.go:775]: I think the partitioning into groups is a complication that we may avoid. Let's discuss vis-a-vis about it.

File: pkg/storegateway/series_chunks.go [hunk 11]
```diff
@@ -331,42 +386,248 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
...
+		partialSeries[i].parsedChunks = nextSet.newSeriesAggrChunkSlice(chunksCount)
+	}
 
-		for j, chunk := range s.chunks {
-			nextSet.series[i].chks[j].MinTime = chunk.minTime
-			nextSet.series[i].chks[j].MaxTime = chunk.maxTime
+	// Create a batched memory pool that can be released all at once. We keep all chunks bytes there.
+	var cachedRanges map[chunkscache.Range][]byte
+	if c.cache != nil {
+		cachedRanges, _ = c.cache.FetchMultiChunks(c.ctx, c.userID, toCacheRanges(nextUnloaded.series, totalGroups), chunksPool)
```
Comment (pracucci) [inline pkg/storegateway/series_chunks.go:408]: [nit] If we don't need `misses` returned by `FetchMultiChunks()` then we can remove it from return value. Maybe we can save some bytes.

**Review by pracucci — COMMENTED** [review]
Massive work! I took a look at the design and logic, skipping tests for now. My main comment is that I think we can't afford to remove the old logic yet. I think we'll need the new logic to be enabled via a feature flag for some time. This is a major change, and will require time to test and rollout to prod. I understand this is out of the of a PoC, but we should think how to make it after the PoC is done and we'll productionise it.

File: pkg/storegateway/chunkscache/memcached.go (hunk 4 above)
Comment (56quarters) [reply pkg/storegateway/chunkscache/memcached.go:29]: yeah, we should use `RemoteCacheClient`. We have two because Reasons

File: pkg/storegateway/chunkscache/memcached.go (hunk 5 above)
Comment (56quarters) [reply pkg/storegateway/chunkscache/memcached.go:4]: I think it's different enough (chunks vs index, addition of pooling) that it counts as new code. @RichiH ?

**Comment by 56quarters** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!
